server/test/emailVerification.spec.js

```diff
@@ -0,0 +1,143 @@
+let hash2;
+const { User } = db;
+
+describe('User SignUp', () => {
+  before((done) => {
+    chai.request(app)
+      .post('/api/users')
+      .send({
+        user: users[5]
+      })
+      .end((err) => {
+        if (err) return done(err);
+        User.findOne({
+          where: {
+            email: users[5].email
+          }
+        }).then((user) => {
+          hash1 = user.hash;
+          done();
+        }).catch(err => done(err));
+      });
+  });
+
+  before((done) => {
+    chai.request(app)
+      .post('/api/users')
+      .send({
+        user: users[6],
+      })
+      .end((err, res) => {
+        if (err) return done(err);
+        hash2 = res.body.user.hash;
+        done();
+      });
+  });
+
+  describe('When passed valid data', () => {
+    it('Should confirm a user email address', (done) => {
+      chai.request(app)
+        .get(`/api/confirmation/${hash1}`)
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(200);
+          expect(res.body.message).to.equal('Your account was successfully activated.');
+          done();
+        });
+    });
+
+    it('Should return an error if user link has expired', (done) => {
+      const func = () => {
+        chai.request(app)
+          .get(`/api/confirmation/${hash2}`)
+          .end((err, res) => {
+            if (err) return done(err);
+            expect(res.status).to.equal(400);
+            expect(res.body.message).to.equal('The verification link has expired.');
+          });
+      };
+      setTimeout(func, 1000000);
+      done();
+    });
+    it('Should return an error if user has already been verified', (done) => {
+      chai.request(app)
+        .get(`/api/confirmation/${hash1}`)
+        .set('Content-Type', 'application/json')
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(400);
+          expect(res.body.message).to.equal('Your account has already been activated.');
+          done();
+        });
+    });
+    it('Should return an error message if user email is not found.', (done) => {
+      chai.request(app)
+        .post('/api/users/reverify')
+        .send({
+          user: {
+            email: 'joetega@gmail.com'
+          }
+        })
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(404);
+          expect(res.body.message).to.equal('The email entered is not registered');
+          done();
+        });
+    });
+    it('Should return an error message if user email is not a valid email.', (done) => {
+      chai.request(app)
+        .post('/api/users/reverify')
+        .send({
+          user: {
+            email: 'joetega.com'
+          }
+        })
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(400);
+          expect(res.body.errors.email[0]).to.equal('Please enter a valid email address.');
```

Comment:
You did a nice job, you can also include in your assertions to test and make sure the user object is not returned when there is validation error, something like `expect(res.body).not.to.have.property('user')

Reply:
Alright, I will do that. Thanks for the feedback.